acme/tf/networks/legal_actions.py

```diff
@@ -0,0 +1,117 @@
+# python3
+# Copyright 2018 DeepMind Technologies Limited. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Networks used for handling illegal actions."""
+
+from typing import Any, Callable, Iterable, Optional, Text, Union
+
+import acme
+from acme.wrappers import open_spiel_wrapper
+import numpy as np
+import sonnet as snt
+import tensorflow as tf
+import tensorflow_probability as tfp
+
+tfd = tfp.distributions
+
+
+class MaskedSequential(snt.Module):
+  """Applies a legal actions mask to a linear chain of modules / callables.
+
+  It is assumed the trailing dimension of the final layer (representing
+  action values) is the same as the trailing dimension of legal_actions.
+  """
+
+  def __init__(self,
+               layers: Iterable[Callable[..., Any]] = None,
+               name: Optional[Text] = None):
+    super(MaskedSequential, self).__init__(name=name)
+    self._layers = list(layers) if layers is not None else []
+    self._illegal_action_penalty = -1e9
+    # TODO Note: illegal_action_penalty cannot be -np.inf, throws
+    # error "Priority must not be NaN"
+
+  def __call__(self, inputs: open_spiel_wrapper.OLT) -> tf.Tensor:
+    # Extract observation, legal actions, and terminal
+    outputs = inputs.observation
+    legal_actions = inputs.legal_actions
+    terminal = inputs.terminal
+
+    for mod in self._layers:
+      outputs = mod(outputs)
+
+    # Apply legal actions mask
+    outputs = tf.where(tf.equal(legal_actions, 1), outputs,
+                       tf.fill(tf.shape(outputs), self._illegal_action_penalty))
+
+    # When computing the Q-learning target (r_t + d_t * max q_t) we need to
+    # ensure max q_t = 0 in terminal states.
+    outputs = outputs * (1 - terminal)
+
+    return outputs
+
+
+# TODO Function to update epsilon
```

Comment:
What does this TODO refer to?

Reply:
Ah good question, I meant to ask...is there currently an easy way to decay DQN's epsilon parameter? There doesn't appear to be, it seems fixed once the policy network is created. This is an important feature for some use cases and since I was building an epsilon greedy sonnet module anyway, it seemed like something we'd want to include.

Reply:
Fixed.